### DISCUSSION TITLE
Why do the setters on MediaRecorderOptions take &mut self?

### DISCUSSION THREAD

**Discussion by allsey87** [wasm-bindgen/wasm-bindgen#3440] [Q&A] [answered]
The methods of MediaRecorderOptions (e.g. [mime_type](https://rustwasm.github.io/wasm-bindgen/api/src/web_sys/features/gen_MediaRecorderOptions.rs.html#60)) take `&mut self` as a receiver. Is this correct and does it actually guarantee exclusive access seeing that MediaRecorderOptions can be cloned?

Link to the WebIDL: https://github.com/rustwasm/wasm-bindgen/blob/main/crates/web-sys/webidls/enabled/MediaRecorder.webidl

**Comment by daxpedda** [comment, marked as answer]
This using a builder pattern is probably not the best API design.

What you have to keep in mind is that `JsValue`s are just references to things that live in the JS runtime. `clone`ing doesn't actually clone the data, it just clones the reference.

You don't really have to be concerned about exclusive access because `MediaRecorderOptions` isn't `Send` or `Sync`, so it can never be called from two places at the same time.

**Reply by allsey87** [reply]
Indeed, I think would be better if the receiver was `&self` though. This makes it clearer that the mutability is handled by the type itself. `&mut self` is misleading in my opinion.

**Reply by daxpedda** [reply]
I really don't know what the original reason to do so was, but as far as I can see this API is a mistake.

I'm not 100% sure, but I think this can still be changed as it shouldn't be a breaking change, if you would like to make a PR that would be great!

**Reply by allsey87** [reply]
I'll have a look into it next week
